File: Graphs/LC_1926_Nearest_Exit_from_Entrance_in_Maze.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.20

class Solution {
public:
    int nearestExit(vector<vector<char>>& maze, vector<int>& entrance) {
        int m = maze.size(), n = maze[0].size();
        queue<pair<int, int>> q;
        int start_row = entrance[0], start_col = entrance[1];
        q.push(make_pair(start_row, start_col));
        maze[start_row][start_col] = '+';
        vector<vector<int>> dirs = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        bool found = false;
        int steps = 0;
        while(!q.empty()){
            int size = q.size();
            for (int k = 0; k < size; k++){
                int cur_row = q.front().first, cur_col = q.front().second;
                q.pop();
                for (int i = 0; i < 4; i++){
                    int nxt_row = cur_row + dirs[i][0], nxt_col = cur_col + dirs[i][1];
                    if (nxt_row >= 0 && nxt_row < m && nxt_col >= 0 && nxt_col < n && maze[nxt_row][nxt_col] == '.'){
                        if (nxt_row == 0 || nxt_row == m-1 || nxt_col == 0 || nxt_col == n-1){
                            return steps + 1;
                        }
                        maze[nxt_row][nxt_col] = '+';
                        q.push(make_pair(nxt_row, nxt_col));
                    } 
                }
            }
            steps++;
        }
        return -1;
    }
};

File: Graphs/LC_1926_Nearest_Exit_from_Entrance_in_Maze.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.20

class Solution:
    def nearestExit(self, maze: List[List[str]], entrance: List[int]) -> int:
        m, n = len(maze), len(maze[0])
        start_row, start_col = entrance[0], entrance[1]
        q = collections.deque()
        q.append((start_row, start_col))
        seen = set()
        seen.add((start_row, start_col))
        steps = 0

        dirs = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        found = 0
        while q:
            size = len(q)
            for _ in range(size):
                cur_row, cur_col = q.popleft()
                if cur_row == 0 or cur_col == 0 or cur_row == m-1 or cur_col == n-1:
                    if cur_row == start_row and cur_col == start_col:
                        found = 0
                    else:
                        found = 1
                if found:
                    return steps
                for dr, dc in dirs:
                    nxt_row, nxt_col = cur_row + dr, cur_col + dc
                    if nxt_row < 0 or nxt_row == m or nxt_col < 0 or nxt_col == n or (nxt_row, nxt_col) in seen or maze[nxt_row][nxt_col] == '+':
                        continue
                    seen.add((nxt_row, nxt_col))
                    q.append((nxt_row, nxt_col))
            steps += 1
        return -1